File: library/src/main/java/com/zeroone/conceal/helper/FileUtils.java
package com.zeroone.conceal.helper;

import android.app.Application;
import android.graphics.Bitmap;
import android.os.Environment;
import android.util.Log;

import com.zeroone.conceal.helper.Constant;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.nio.channels.FileChannel;

import static com.zeroone.conceal.helper.Constant.DEFAULT_DIRECTORY;
import static com.zeroone.conceal.helper.Constant.DEFAULT_IMAGE_FOLDER;

/**
 * @author : hafiq on 24/03/2017.
 */

public class FileUtils {

    public static File getImageDirectory(String mFolderName){
        File file = new File(DEFAULT_DIRECTORY+mFolderName+"/"+DEFAULT_IMAGE_FOLDER);
        Log.d("Conceal",file.getAbsolutePath());
        if (file.mkdirs())
            return file;
        if (file.exists())
            return file;

        return null;
    }

    public static File makeFile(File dir,String filename){
        File file = new File(dir,filename);
        try {
            if (file.exists())
                file.delete();

            if (file.createNewFile()){
                return file;
            }
        } catch (IOException e) {
            e.printStackTrace();
            return null;
        }

        return null;
    }

    public static String getExtension(File file){
        return file.getAbsolutePath().substring(file.getAbsolutePath().lastIndexOf("."));
    }

    public static String getExtension(String file){
        return file.substring(file.lastIndexOf("."));
    }


    public static File moveFile(File file, File dir){
        if (!dir.exists())
            dir.mkdirs();

        File newFile = new File(dir, file.getName());
        FileChannel outputChannel;
        FileChannel inputChannel;
        try {
            outputChannel = new FileOutputStream(newFile).getChannel();
            inputChannel = new FileInputStream(file).getChannel();
            inputChannel.transferTo(0, inputChannel.size(), outputChannel);
            inputChannel.close();

            inputChannel.close();
            outputChannel.close();

            file.delete();

            Log.d("Files","File have been moved to : "+newFile.getAbsolutePath());
        }
        catch (Exception e){
            return null;
        }

        return newFile;
    }

    public static byte[] readContentIntoByteArray(File file)
    {
        FileInputStream fileInputStream = null;
        byte[] bFile = new byte[(int) file.length()];
        try
        {
            //convert file into array of bytes
            fileInputStream = new FileInputStream(file);
            fileInputStream.read(bFile);
            fileInputStream.close();
            for (int i = 0; i < bFile.length; i++)
            {
                System.out.print((char) bFile[i]);
            }
        }
        catch (Exception e)
        {
            e.printStackTrace();
        }
        return bFile;
    }


    public static boolean saveBitmap(File imageFile, Bitmap bitmap) {

        boolean fileCreated = false;
        boolean bitmapCompressed = false;
        boolean streamClosed = false;

        if (imageFile.exists())
            if (!imageFile.delete())
                return false;

        try {
            fileCreated = imageFile.createNewFile();
        } catch (IOException e) {
            e.printStackTrace();
        }

        FileOutputStream out = null;
        try {
            out = new FileOutputStream(imageFile);
            bitmapCompressed = bitmap.compress(Bitmap.CompressFormat.PNG, 100, out);

        } catch (Exception e) {
            e.printStackTrace();
            bitmapCompressed = false;

        } finally {
            if (out != null) {
                try {
                    out.flush();
                    out.close();
                    streamClosed = true;

                } catch (IOException e) {
                    e.printStackTrace();
                    streamClosed = false;
                }
            }
        }

        return (fileCreated && bitmapCompressed && streamClosed);
    }

}

File: library/src/main/java/com/chamber/kotlin/library/FileUtils.kt
package com.chamber.kotlin.library

import android.graphics.Bitmap
import com.chamber.kotlin.library.model.Constant.DEFAULT_DIRECTORY
import com.chamber.kotlin.library.model.Constant.DEFAULT_IMAGE_FOLDER
import com.chamber.kotlin.library.model.Constant.DEFAULT_PREFIX_FILENAME
import com.chamber.kotlin.library.model.CryptoFile
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException
import java.nio.channels.FileChannel
import java.util.*

/**
 * @author : hafiq on 24/03/2017.
 */

internal object FileUtils {
    @JvmStatic
    fun moveFile(file: File?, dir: File?): File? {
        if (dir == null || file == null)
            return null

        if (!dir.exists()) {
            if (dir.mkdirs()) {
                val newFile = File(dir, file.name)
                val outputChannel: FileChannel
                val inputChannel: FileChannel
                try {
                    outputChannel = FileOutputStream(newFile).channel
                    inputChannel = FileInputStream(file).channel
                    inputChannel.transferTo(0, inputChannel.size(), outputChannel)
                    inputChannel.close()

                    inputChannel.close()
                    outputChannel.close()

                    file.delete()
                } catch (e: Exception) {
                    return null
                }

                return newFile
            }
        }

        return file
    }

    @JvmStatic
    fun saveBitmap(imageFile: File, bitmap: Bitmap): Boolean {

        var fileCreated = false
        var bitmapCompressed: Boolean
        var streamClosed = false

        if (imageFile.exists())
            if (!imageFile.delete())
                return false

        try {
            fileCreated = imageFile.createNewFile()
        } catch (e: IOException) {
            e.printStackTrace()
        }

        var out: FileOutputStream? = null
        try {
            out = FileOutputStream(imageFile)
            bitmapCompressed = bitmap.compress(Bitmap.CompressFormat.PNG, 100, out)

        } catch (e: Exception) {
            e.printStackTrace()
            bitmapCompressed = false

        } finally {
            if (out != null) {
                try {
                    out.flush()
                    out.close()
                    streamClosed = true

                } catch (e: IOException) {
                    e.printStackTrace()
                    streamClosed = false
                }

            }
        }

        return fileCreated && bitmapCompressed && streamClosed
    }

    @JvmStatic
    fun isFileForImage(file: File?): Boolean {
        if (file == null)
            return false

        val okFileExtensions = arrayOf("jpg", "png", "gif", "jpeg")

        for (extension in okFileExtensions) {
            if (file.name.toLowerCase().endsWith(extension)) {
                return true
            }
        }
        return false
    }

    /***
     * get default directory
     * @return File
     */
    @JvmStatic
    fun getDirectory(mFolderName: String): File? {
        val file = File(DEFAULT_DIRECTORY + mFolderName + "/" + DEFAULT_IMAGE_FOLDER)
        return if (file.exists()) file else null

    }

    /***
     * get default folder
     * @return File
     */
    @JvmStatic
    fun getImageDirectory(mFolderName: String): File? {
        val file = File(DEFAULT_DIRECTORY + mFolderName + "/" + DEFAULT_IMAGE_FOLDER)
        if (file.mkdirs())
            return file
        return if (file.exists()) file else null

    }

    /***
     * get List of encrypted file
     * @param parentDir - root directory
     * @return File
     */
    @JvmStatic
    fun getListFiles(parentDir: File?): List<CryptoFile> {
        val inFiles = ArrayList<CryptoFile>()
        try {
            if (parentDir != null) {
                val files = parentDir.listFiles()
                for (file in files!!) {
                    if (file.isDirectory) {
                        inFiles.addAll(getListFiles(file))
                    } else {
                        if (file.name.startsWith(DEFAULT_PREFIX_FILENAME)) {
                            val cryptoFile = CryptoFile()
                            cryptoFile.fileName = file.name
                            cryptoFile.path = file.absolutePath
                            cryptoFile.type = file.parent
                            inFiles.add(cryptoFile)
                        }
                    }
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }

        return inFiles
    }
}
